### PR TITLE
Secure REST endpoints with Consoledot RBAC

### PR DESCRIPTION
**Pull request by ebaron**
This is a port of https://github.com/RedHatInsights/notifications-backend/commit/2177c7ce159cabc97be5c86aa8c4dc3dc26436f4, modified to work with runtimes-inventory. This integrates with [Quarkus Security architecture](https://quarkus.io/guides/security-architecture), which allows us to drop in the RBAC integration with very little modification to our existing endpoints.

Some key changes from the original notifications-backend implementation:
- Using the Host Inventory's RBAC permissions, e.g. `inventory:hosts:[read|write]`
- Removing [Turnpike](https://github.com/RedHatInsights/turnpike) support. The rest of runtimes-inventory is set up to expect an `x-rh-identity` header with an org ID from 3scale. So I chose to simplify by removing Turnpike-specific code for now, and focus solely on authentication coming from 3scale.
- Removed code handling internal APIs, since we currently have none.
- Removed code which queries groups from the RBAC server. This is used by a particular notifications-backend endpoint, and doesn't appear needed for us.

I tested this in an Ephemeral Environment with the default user, and a custom user who has an empty permission set.

The first case with the default user succeeds as expected:
```
curl -sSf -H "Authorization: Basic $BASIC_AUTH" https://ephemeral.example.com/api/runtimes-inventory-service/v1/instances/
{"response": "[]"}
```

The second case with an unprivileged user fails as expected:
```
curl -sSf -H "Authorization: Basic $BASIC_AUTH_OTHER" https://ephemeral.example.com/api/runtimes-inventory-service/v1/instances/
curl: (22) The requested URL returned error: 403 
```

The pod logs show both access attempts:
```
2023-09-25 20:23:20,474 INFO [access_log] (executor-thread-2) 127.0.0.1 - defaultuser 25/Sep/2023:20:23:20 +0000 "GET /api/runtimes-inventory-service/v1/instances/ HTTP/1.1" 200 18 "-" "curl/8.0.1"
2023-09-25 20:24:43,718 INFO [access_log] (vert.x-eventloop-thread-1) 127.0.0.1 - myuser 25/Sep/2023:20:24:43 +0000 "GET /api/runtimes-inventory-service/v1/instances/ HTTP/1.1" 403 - "-" "curl/8.0.1"
```

### CODE REVIEW

File: core/src/main/java/com/redhat/runtimes/inventory/auth/ConsoleIdentityProvider.java [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+                    // A failure will cause an authentication failure
+                    .onFailure()
+                    .transform(
+                        throwable -> {
+                          Log.error("Error while processing identity", throwable);
+                          return new AuthenticationFailedException(throwable);
+                        }));
+  }
+
+  private static ConsoleIdentity getRhIdentityFromString(String xRhIdHeader) {
```
Comment (kittylyst) [inline core/src/main/java/com/redhat/runtimes/inventory/auth/ConsoleIdentityProvider.java:161]: Can this be made default access & unit tested?

File: core/src/main/java/com/redhat/runtimes/inventory/auth/principal/IllegalIdentityHeaderException.java [hunk 2]
```diff
@@ -0,0 +1,10 @@
+/* Copyright (C) Red Hat 2023 */
+package com.redhat.runtimes.inventory.auth.principal;
+
+// Derived from notifications-backend
+public class IllegalIdentityHeaderException extends Exception {
+
```
Comment (kittylyst) [inline core/src/main/java/com/redhat/runtimes/inventory/auth/principal/IllegalIdentityHeaderException.java:6]: Custom exception classes should define all constructors.

File: core/src/main/java/com/redhat/runtimes/inventory/auth/rbac/RbacClientResponseFilter.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+public class RbacClientResponseFilter implements ClientResponseFilter {
+
+  @Override
+  public void filter(ClientRequestContext requestContext, ClientResponseContext responseContext)
+      throws IOException {
+    Response.StatusType statusInfo = responseContext.getStatusInfo();
+    int status = statusInfo.getStatusCode();
+    if (status == 0) {
+      Log.infof(
+          "Call to the Rbac server failed with code %d, %s", status, statusInfo.getReasonPhrase());
```
Comment (kittylyst) [inline core/src/main/java/com/redhat/runtimes/inventory/auth/rbac/RbacClientResponseFilter.java:26]: Doesn't the case 0 here indicate a failure to respond at all?

File: core/src/main/java/com/redhat/runtimes/inventory/auth/rbac/RbacRaw.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+          if (fields[2].equals(what) || fields[2].equals("*")) {
+            return true;
+          }
+        }
+      }
+    }
+    return false;
+  }
+
+  private String[] getPermissionFields(Map<String, Object> map) {
```
Comment (kittylyst) [inline core/src/main/java/com/redhat/runtimes/inventory/auth/rbac/RbacRaw.java:46]: Consider inlining this method?

File: core/src/main/java/com/redhat/runtimes/inventory/auth/rbac/RbacRestClientRequestFilter.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+import java.io.IOException;
+import java.net.URI;
+import java.util.Base64;
+
+// Derived from notifications-backend
+
+/**
+ * Filter to optionally add data on outgoing requests to the RBAC service. This is meant for local
+ * development and not production.
+ */
```
Comment (kittylyst) [inline core/src/main/java/com/redhat/runtimes/inventory/auth/rbac/RbacRestClientRequestFilter.java:17]: Are we sure that we want this leaking into code that's going to get deployed to prod? Are there any credible alternatives?

File: core/src/test/java/com/redhat/runtimes/inventory/auth/rbac/RbacParsingTest.java [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+class RbacParsingTest {
+
+  private static final ObjectReader RBAC_RAW_READER = new ObjectMapper().readerFor(RbacRaw.class);
+  private static final String BASE = "src/test/resources/rbac-examples/";
+  public static final String ALL = "*";
+
+  @Test
+  void testParseExample() throws Exception {
+    RbacRaw rbac = readValueFromFile("rbac_example.json");
+    assert rbac.data.size() == 2;
```
Comment (kittylyst) [inline core/src/test/java/com/redhat/runtimes/inventory/auth/rbac/RbacParsingTest.java:22]: assert is a keyword, not the JUnit 5 methods. Please remove all uses of the assert keyword & replace with appropriate JUnit 5 assertions.

**Review by kittylyst — CHANGES_REQUESTED** [review]
Not bad - a few minor things to clean up.

One important thing: All uses of the assert keyword must be replaced with appropriate JUnit 5 assertions. (Ask me in Slack if you're interested in this gory and rather sad story).

File: core/src/main/java/com/redhat/runtimes/inventory/auth/ConsoleAuthMechanism.java [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+  @Override
+  public Set<Class<? extends AuthenticationRequest>> getCredentialTypes() {
+    return Collections.singleton(ConsoleAuthenticationRequest.class);
+  }
+
+  @Override
+  public HttpCredentialTransport getCredentialTransport() {
+    return null;
+  }
```
[the remark this answers is not in the document]
Comment (ebaron) [reply core/src/main/java/com/redhat/runtimes/inventory/auth/ConsoleAuthMechanism.java:80]: I believe so. The docs say the following:
> The credential transport, used for finding the best candidate for authenticating and challenging when more than one mechanism is installed. and finding the best candidate for issuing a challenge when more than one mechanism is installed. May be null if this mechanism cannot interfere with other mechanisms

In this case there is only the one mechanism installed. This is what's used in notifications-backend.

File: core/src/main/java/com/redhat/runtimes/inventory/auth/rbac/RbacClientResponseFilter.java (hunk 3 above)
Comment (ebaron) [reply core/src/main/java/com/redhat/runtimes/inventory/auth/rbac/RbacClientResponseFilter.java:26]: This is meant to reduce warnings logged when the RBAC server times out: https://github.com/RedHatInsights/notifications-backend/pull/1322

File: core/src/main/java/com/redhat/runtimes/inventory/auth/rbac/RbacRestClientRequestFilter.java (hunk 5 above)
Comment (ebaron) [reply core/src/main/java/com/redhat/runtimes/inventory/auth/rbac/RbacRestClientRequestFilter.java:17]: It doesn't serve a purpose for us currently, so I'll just remove it.

File: core/src/main/java/com/redhat/runtimes/inventory/auth/ConsoleIdentityProvider.java (hunk 1 above)
Comment (ebaron) [reply core/src/main/java/com/redhat/runtimes/inventory/auth/ConsoleIdentityProvider.java]: Added tests under com.redhat.runtimes.inventory.auth.RhIdentityParsingTest using some example RhId headers from notifications-backend.

File: core/src/main/java/com/redhat/runtimes/inventory/auth/principal/IllegalIdentityHeaderException.java (hunk 2 above)
Comment (ebaron) [reply core/src/main/java/com/redhat/runtimes/inventory/auth/principal/IllegalIdentityHeaderException.java:6]: Done

File: core/src/main/java/com/redhat/runtimes/inventory/auth/rbac/RbacRaw.java (hunk 4 above)
Comment (ebaron) [reply core/src/main/java/com/redhat/runtimes/inventory/auth/rbac/RbacRaw.java]: Sure

File: core/src/test/java/com/redhat/runtimes/inventory/auth/rbac/RbacParsingTest.java (hunk 6 above)
Comment (ebaron) [reply core/src/test/java/com/redhat/runtimes/inventory/auth/rbac/RbacParsingTest.java]: Replaced all asserts with JUnit 5 assertions

**Review by kittylyst — APPROVED** [review]
Approved, but please don't merge yet - want to make sure the Hibernate work lands first.

**Pull request merged**